modvendor/modvendor.go

```diff
@@ -15,59 +15,220 @@
 package modvendor
 
 import (
+	"fmt"
+	iofs "io/fs"
+	"io/ioutil"
 	"os"
 	"path/filepath"
+	"sort"
+	"strings"
 
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclwrite"
 	"github.com/mineiros-io/terramate/errors"
 	"github.com/mineiros-io/terramate/fs"
 	"github.com/mineiros-io/terramate/git"
 	"github.com/mineiros-io/terramate/tf"
 	"github.com/rs/zerolog/log"
+	"github.com/zclconf/go-cty/cty"
 )
 
 const (
 	// ErrAlreadyVendored indicates that a module is already vendored.
 	ErrAlreadyVendored errors.Kind = "module is already vendored"
 )
 
-// Vendor will vendor the given module inside the provided vendor dir.
+type modinfo struct {
+	source     string
+	vendoredAt string
+	origin     string
+}
+
+// Vendor will vendor the given module and its dependencies inside the provided
+// root dir.
+// The root dir must be an absolute path.
 // The vendor dir must be an absolute path that will be considered as relative
 // to the given rootdir.
 //
 // Vendored modules will be located at:
 //
 // - <rootdir>/<vendordir>/<Source.Path>/<Source.Ref>
 //
-// If the project is already vendored an error of kind ErrAlreadyVendored will
-// be returned, vendored projects are never updated.
 //
 // The whole path inside the vendor dir will be created if it not exists.
-// Vendoring is not recursive, so dependencies won't have their dependencies vendored.
-// Vendoring will also not download any git submodules.
+// Vendoring will not download any git submodules.
+//
+// The remote git module dependencies will also be vendored and each
+// module.source declaration for those dependencies will be rewritten to
+// reference them inside the vendor directory.
 //
-// It returns the absolute path where the module has been vendored.
-func Vendor(rootdir, vendordir string, modsrc tf.Source) (string, error) {
+// It returns a report of everything vendored and ignored (with a reason).
+func Vendor(rootdir string, vendorDir string, modsrc tf.Source) Report {
+	report := NewEmptyReport()
+	if !filepath.IsAbs(vendorDir) {
+		report.Error = errors.E("vendor dir %q must be absolute path", vendorDir)
+		return report
+	}
+
+	return recVendor(rootdir, vendorDir, modsrc, report, nil)
+}
+
+// VendorAll will vendor all dependencies of the tfdir into rootdir.
+// It will scan all .tf files in the directory and vendor each module declaration
+// containing the supported remote source URLs.
+func VendorAll(rootdir string, vendorDir string, tfdir string) Report {
+	return vendorAll(rootdir, vendorDir, tfdir, NewEmptyReport())
+}
+
+func recVendor(rootdir string, vendorDir string, modsrc tf.Source, report Report, info *modinfo) Report {
+	logger := log.With().
+		Str("action", "modvendor.recVendor()").
+		Str("module.source", modsrc.Raw).
+		Logger()
+
+	moddir, err := doVendor(rootdir, vendorDir, modsrc)
```

Comment:
hmm `recVendor` + `doVendor` seems confusing. Usually when I have a recursive implementation the name of the function is the same but private, like `vendor`, so you have Vendor calling an internal recursive vendor() function. And then we would need a better name to doVendor, not sure yet, maybe `downloadVendor`.